src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Enter.java

```diff
@@ -439,7 +439,7 @@ public void visitClassDef(JCClassDecl tree) {
                 log.error(tree.pos(),
                           Errors.ClassPublicShouldBeInFile(topElement, tree.name));
             }
-            if ((tree.mods.flags & UNNAMED_CLASS) != 0) {
+            if ((tree.mods.flags & IMPLICIT_CLASS) != 0) {
```

Comment:
I'm not 100% sure about this. Another option would be to tweak Resolve::findType and make sure that `typeNotFound` (or some more specific message!) is returned when an implicit class is referred to by name. If you look at the first few lines in that method you will see:

```
if (name == names.empty)
            return typeNotFound; // do not allow inadvertent "lookup" of anonymous types
```

So, there's already some logic there for skipping some types.

Reply:
The main thing here is that it is done once at the source instead of every look up. The symbol should not exist in the compilation unit.

Reply:
Yep - I think the main consequence of this choice is the fact that we cannot generate a specific error message - but perhaps that's ok (after all, having a special message would equate to admitting that the compiler _does know_ about the name :-) )